<a href="https://colab.research.google.com/github/Pranavla/Breast_Cancer_Prediction_Using_ANN/blob/main/Breast_Cancer_Prediction_Using_ANN.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Importing necessary libraries**

In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

# **Importing the dataset**

In [ ]:
df=pd.read_csv('/content/breast_cancer.csv')

Printing the first 10 rows in the dataset

In [ ]:
df.head(10)

id diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
0    842302         M        17.99         10.38          122.80     1001.0   
1    842517         M        20.57         17.77          132.90     1326.0   
2  84300903         M        19.69         21.25          130.00     1203.0   
3  84348301         M        11.42         20.38           77.58      386.1   
4  84358402         M        20.29         14.34          135.10     1297.0   
5    843786         M        12.45         15.70           82.57      477.1   
6    844359         M        18.25         19.98          119.60     1040.0   
7  84458202         M        13.71         20.83           90.20      577.9   
8    844981         M        13.00         21.82           87.50      519.8   
9  84501001         M        12.46         24.04           83.97      475.9   

   smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
0          0.11840           0.27760         0.30010              0.14710   
1          0.08474           0.07864         0.08690              0.07017   
2          0.10960           0.15990         0.19740              0.12790   
3          0.14250           0.28390         0.24140              0.10520   
4          0.10030           0.13280         0.19800              0.10430   
5          0.12780           0.17000         0.15780              0.08089   
6          0.09463           0.10900         0.11270              0.07400   
7          0.11890           0.16450         0.09366              0.05985   
8          0.12730           0.19320         0.18590              0.09353   
9          0.11860           0.23960         0.22730              0.08543   

   ...  texture_worst  perimeter_worst  area_worst  smoothness_worst  \
0  ...          17.33           184.60      2019.0            0.1622   
1  ...          23.41           158.80      1956.0            0.1238   
2  ...          25.53           152.50      1709.0            0.1444   
3  ...          26.50            98.87       567.7            0.2098   
4  ...          16.67           152.20      1575.0            0.1374   
5  ...          23.75           103.40       741.6            0.1791   
6  ...          27.66           153.20      1606.0            0.1442   
7  ...          28.14           110.60       897.0            0.1654   
8  ...          30.73           106.20       739.3            0.1703   
9  ...          40.68            97.65       711.4            0.1853   

   compactness_worst  concavity_worst  concave points_worst  symmetry_worst  \
0             0.6656           0.7119                0.2654          0.4601   
1             0.1866           0.2416                0.1860          0.2750   
2             0.4245           0.4504                0.2430          0.3613   
3             0.8663           0.6869                0.2575          0.6638   
4             0.2050           0.4000                0.1625          0.2364   
5             0.5249           0.5355                0.1741          0.3985   
6             0.2576           0.3784                0.1932          0.3063   
7             0.3682           0.2678                0.1556          0.3196   
8             0.5401           0.5390                0.2060          0.4378   
9             1.0580           1.1050                0.2210          0.4366   

   fractal_dimension_worst  Unnamed: 32  
0                  0.11890          NaN  
1                  0.08902          NaN  
2                  0.08758          NaN  
3                  0.17300          NaN  
4                  0.07678          NaN  
5                  0.12440          NaN  
6                  0.08368          NaN  
7                  0.11510          NaN  
8                  0.10720          NaN  
9                  0.20750          NaN  

[10 rows x 33 columns]

Coiunting the number of rows and columns in the dataset

In [ ]:
df.shape

(569, 33)

From here we can understand that the dataset is having 569 rows and 33 columns .The dataset is a quality dataset it not having any missing values or unregular blocks . So, we are seperating X(Input) and y(output). The diagnosis column is the output column and other values can be taken for input.The id column is not required.

Printing the column names

In [ ]:
print(df.columns)

Index(['id', 'diagnosis', 'radius_mean', 'texture_mean', 'perimeter_mean',
       'area_mean', 'smoothness_mean', 'compactness_mean', 'concavity_mean',
       'concave points_mean', 'symmetry_mean', 'fractal_dimension_mean',
       'radius_se', 'texture_se', 'perimeter_se', 'area_se', 'smoothness_se',
       'compactness_se', 'concavity_se', 'concave points_se', 'symmetry_se',
       'fractal_dimension_se', 'radius_worst', 'texture_worst',
       'perimeter_worst', 'area_worst', 'smoothness_worst',
       'compactness_worst', 'concavity_worst', 'concave points_worst',
       'symmetry_worst', 'fractal_dimension_worst', 'Unnamed: 32'],
      dtype='object')


In [ ]:
X = df.iloc[:,2:32].values
y = df.iloc[:,1].values

We need to transform the categorical data into numerical data so am using label encoding for it.

In [ ]:
from sklearn.preprocessing import LabelEncoder
labelencoder = LabelEncoder()

In [ ]:
y = labelencoder.fit_transform(y)

In [ ]:
y

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1,
       1, 1, 0, 1, 0, 0, 0, 0, 0, 1, 1, 0, 1, 1, 0, 0, 0, 0, 1, 0, 1, 1,
       0, 0, 0, 0, 1, 0, 1, 1, 0, 1, 0, 1, 1, 0, 0, 0, 1, 1, 0, 1, 1, 1,
       0, 0, 0, 1, 0, 0, 1, 1, 0, 0, 0, 1, 1, 0, 0, 0, 0, 1, 0, 0, 1, 0,
       0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 1, 1, 0, 0, 0, 1, 1, 0, 1, 0, 1,
       1, 0, 1, 1, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 1, 0, 0, 0, 0, 1, 1, 0, 1, 0, 0, 1, 1, 0, 0, 1, 1, 0, 0, 0,
       0, 1, 0, 0, 1, 1, 1, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 1, 1, 0, 1, 1,
       1, 1, 0, 1, 1, 1, 0, 1, 0, 1, 0, 0, 1, 0, 1, 1, 1, 1, 0, 0, 1, 1,
       0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 1, 0, 0, 1, 0, 0, 1, 1, 0, 1, 0, 0,
       0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 1, 0, 0, 1, 0, 1, 1, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1,

# **Training**

Seperating training and testing data and scaling the data

In [ ]:
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test = train_test_split(X,y,test_size = 0.3,random_state = 42)

In [ ]:
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.fit_transform(X_test)

# **ANN**

We need keras package for training the neural network

In [ ]:
!pip install -q keras

In [ ]:
import keras
from keras.models import Sequential
from keras.layers import Dense

## Initialising ANN

In [ ]:
classifier = Sequential()

## Creating input and first hidden layer

In [ ]:
classifier.add(Dense(units = 16,kernel_initializer = 'uniform',activation = 'sigmoid',input_dim = 30))

## Creating Second hidden layer

In [ ]:
classifier.add(Dense(units = 16,kernel_initializer = 'uniform',activation='sigmoid'))

## Creating output layer

In [ ]:
classifier.add(Dense(units = 1,kernel_initializer = 'uniform',activation='relu'))

# **Training the dataset**

In [ ]:
classifier.compile(optimizer = 'adam',loss = 'binary_crossentropy',metrics = ['accuracy'])

In [ ]:
classifier.fit(X_train,y_train,batch_size = 10,epochs = 100)

Epoch 1/100
40/40 [==============================] - 2s 5ms/step - loss: 0.7332 - accuracy: 0.6256
Epoch 2/100
40/40 [==============================] - 0s 5ms/step - loss: 0.6418 - accuracy: 0.6256
Epoch 3/100
40/40 [==============================] - 0s 5ms/step - loss: 0.6088 - accuracy: 0.6256
Epoch 4/100
40/40 [==============================] - 0s 5ms/step - loss: 0.5683 - accuracy: 0.6482
Epoch 5/100
40/40 [==============================] - 0s 5ms/step - loss: 0.5245 - accuracy: 0.8844
Epoch 6/100
40/40 [==============================] - 0s 5ms/step - loss: 0.4747 - accuracy: 0.9322
Epoch 7/100
40/40 [==============================] - 0s 5ms/step - loss: 0.4132 - accuracy: 0.9472
Epoch 8/100
40/40 [==============================] - 0s 5ms/step - loss: 0.3396 - accuracy: 0.9548
Epoch 9/100
40/40 [==============================] - 0s 5ms/step - loss: 0.2569 - accuracy: 0.9598
Epoch 10/100
40/40 [==============================] - 0s 3ms/step - loss: 0.1714 - accuracy: 0.9648
Epoch 11/

In [ ]:
from sklearn.metrics import accuracy_score
y_pred = classifier.predict(X_test)
y_pred = (y_pred>0.5)
accuracy = accuracy_score(y_test,y_pred)
print("Accuracy :%.2f%% " %(accuracy*100))

6/6 [==============================] - 0s 4ms/step
Accuracy :98.83% 


From this we can understand that the model is **having an accuracy of 99%** which is amazing

Let's print the confusion metrics to get the positives and negatives

In [ ]:
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(y_test,y_pred)

In [ ]:
cm

array([[107,   1],
       [  1,  62]])

## **Testing the perfomance of the model**

In [ ]:
print("Enter the following values to predict breast cancer diagnosis: 1.Radius_mean, 2.Texture_mean, 3.Perimeter_mean, 4.Area_mean, 5.Smoothness_mean, 6.compactness_mean,7.Concavity_mean,8.Cconcave points_mean, 9.Symmetry_mean, 10.Fractal_dimension_mean',11.Radius_se, 12.Texture_se, 13.Perimeter_se, 14.Area_se, 15.Smoothness_se,16.Compactness_se', 17.Concavity_se, 18.Concave points_se, 19.symmetry_se,20.Fractal_dimension_se, 21.Radius_worst', 22.Texture_worst,23.Perimeter_worst', 24.Area_worst, 25.smoothness_worst, 26.Compactness_worst, 27.Concavity_worst, 28.Concave points_worst', 29.Symmetry_worst, 30.Fractal_dimension_worst00.2871")
input_data = []
for i in range(30):
    value = float(input(f"Enter value {i+1}: "))
    input_data.append(value)
new_data = np.array([input_data])
new_data_scaled = sc.transform(new_data)
prediction = classifier.predict(new_data_scaled)
if prediction > 0.5:
    diagnosis = "Malignant"
else:
    diagnosis = "Benign"
print("Prediction for input data:", diagnosis)

Enter the following values to predict breast cancer diagnosis: 1.Radius_mean, 2.Texture_mean, 3.Perimeter_mean, 4.Area_mean, 5.Smoothness_mean, 6.compactness_mean,7.Concavity_mean,8.Cconcave points_mean, 9.Symmetry_mean, 10.Fractal_dimension_mean',11.Radius_se, 12.Texture_se, 13.Perimeter_se, 14.Area_se, 15.Smoothness_se,16.Compactness_se', 17.Concavity_se, 18.Concave points_se, 19.symmetry_se,20.Fractal_dimension_se, 21.Radius_worst', 22.Texture_worst,23.Perimeter_worst', 24.Area_worst, 25.smoothness_worst, 26.Compactness_worst, 27.Concavity_worst, 28.Concave points_worst', 29.Symmetry_worst, 30.Fractal_dimension_worst00.2871
Enter value 1: 20.6
Enter value 2: 29.33
Enter value 3: 140.1
Enter value 4: 1265
Enter value 5: 0.1178
Enter value 6: 0.277
Enter value 7: 0.3514
Enter value 8: 0.152
Enter value 9: 0.2397
Enter value 10: 0.07016
Enter value 11: 0.726
Enter value 12: 1.595
Enter value 13: 5.772
Enter value 14: 86.22
Enter value 15: 0.006522
Enter value 16: 0.06158
Enter value 17

# **Conclusion**

Here I have created a model to predict breast cancer by using ANN.I have created a detailed article on ANN and it's implementation((https://medium.com/@pranavlal568/unveiling-the-power-of-artificial-neural-networks-implementation-and-future-directions-5c771977e856) It is having an accuracy of **98.83%** which is phenominal for a model for breast  cancer prediction . Here the term malignant refers to person having cancerous cell and benign refers to person not having the cancerous tissue.